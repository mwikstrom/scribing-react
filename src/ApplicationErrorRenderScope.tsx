import React, { createContext, ReactNode, useContext } from "react";

/** @public */
export interface ApplicationErrorRenderScopeProps {
    renderErrorInfo: (error: Error) => ReactNode;
    children: ReactNode;
}

/** @public */
export function ApplicationErrorRenderScope(props: ApplicationErrorRenderScopeProps): JSX.Element {
    const { renderErrorInfo, ...rest } = props;
    return <RenderErrorInfoContext.Provider value={renderErrorInfo} {...rest}/>;
}

/** @public */
export function useApplicationErrorRenderer(): ApplicationErrorRenderScopeProps["renderErrorInfo"] | null {
    return useContext(RenderErrorInfoContext);
}

const RenderErrorInfoContext = createContext<ApplicationErrorRenderScopeProps["renderErrorInfo"] | null>(null);
