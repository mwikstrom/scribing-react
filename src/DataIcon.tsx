import React, { CSSProperties, FC, useMemo } from "react";
import Icon from "@mdi/react";
import { mdiAlertOctagonOutline, mdiAlertOutline, mdiCheckCircleOutline, mdiInformationOutline } from "@mdi/js";
import { PredefinedIcon, PREDEFINED_ICONS } from "scribing";
import { 
    ICON_PACKS, 
    useMaterialDesignIconPath, 
    useMaterialDesignIcons, 
    useMaterialDesignIconTags 
} from "./internal/IconPack";

/** @public */
export interface DataIconProps {
    className?: string;
    data?: string;
}

/** @public */
export const DataIcon: FC<DataIconProps> = props => {
    const { data = "" } = props;
    if (MDI_PATTERN.test(data)) {
        return <MdiIcon {...props}/>;
    } else {
        return <PredefinedOrPathIcon {...props}/>;
    }
};

/** @public */
export function useDataIconPacks(): readonly string[] {
    return ICON_PACKS;
}

/** @public */
export function useDataIconTags(pack: string): readonly string[] | null {
    const mdiTags = useMaterialDesignIconTags();
    if (pack === "mdi") {
        return mdiTags;
    } else {
        return null;
    }
}

/** @public */
export function useDataIcons(pack?: string, tag?: string): readonly string[] {
    const mdiIcons = useMaterialDesignIcons(tag);
    return useMemo(() => {
        const result: string[] = [];
        if (!pack || pack === "predefined") {
            result.push(...PREDEFINED_ICONS);
        }
        if ((!pack || pack === "mdi") && mdiIcons) {
            result.push(...mdiIcons);
        }
        return result;
    }, [pack, mdiIcons]);
}

const MdiIcon: FC<DataIconProps> = ({className, data = ""}) => {
    const iconName = data.replace(MDI_PATTERN, "");
    const path = useMaterialDesignIconPath(iconName);
    const iconStyle = useMemo<CSSProperties>(() => ({
        transition: "opacity ease-out 0.2s",
        opacity: path ? 1 : 0,
        userSelect: "none",
    }), [path]);
    return (<>&#x200B;<Icon path={path} className={className} style={iconStyle}/></>);
};

const PredefinedOrPathIcon: FC<DataIconProps> = ({className, data = ""}) => {
    const path = useMemo(() => {
        if (data in PREDEFINED_ICON_PATHS) {
            return PREDEFINED_ICON_PATHS[data as PredefinedIcon];
        } else if (ICON_NAME_PATTERN.test(data)) {
            console.warn("Unsupported icon: ", data);
            return "";
        } else {
            return data;
        }
    }, [data]);

    const iconStyle = useMemo<CSSProperties | undefined>(() => {
        if (!path) {
            return {
                backgroundImage: `repeating-linear-gradient(
                    -45deg,
                    currentcolor 0px,
                    currentcolor 3px,
                    transparent 3px,
                    transparent 6px
                )`,        
            };
        }
    }, [path]);

    return (<Icon path={path} className={className} style={iconStyle}/>);
};

const MDI_PATTERN = /^@mdi\//;
const ICON_NAME_PATTERN = /^(?:@[a-z-]+\/)?[a-z-]+$/i;

const PREDEFINED_ICON_PATHS: Readonly<Record<PredefinedIcon, string>> = Object.freeze({
    information: mdiInformationOutline,
    success: mdiCheckCircleOutline,
    warning: mdiAlertOutline,
    error: mdiAlertOctagonOutline,
});
