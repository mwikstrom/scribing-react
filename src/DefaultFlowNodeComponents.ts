import { ParagraphStyleVariant } from "scribing";
import { DynamicTextView } from "./DynamicTextView";
import { FlowButtonView } from "./FlowButtonView";
import { FlowNodeComponentMap, ParagraphComponent } from "./FlowNodeComponent";
import { LineBreakView } from "./LineBreakView";
import { ParagraphBreakView } from "./ParagraphBreakView";
import { TextRunView } from "./TextRunView";
import { UnknownNodeView } from "./UnknownNodeView";

const getParagraphComponent = (variant: ParagraphStyleVariant): ParagraphComponent => {
    switch (variant) {
    case "h1":
    case "h2":
    case "h3":
    case "h4":
    case "h5":
    case "h6":
        return variant;
    
    default:
        return "p";
    }
};

/** @public */
export const DefaultFlowNodeComponents: Readonly<FlowNodeComponentMap> = Object.freeze({
    textRunView: TextRunView,
    lineBreakView: LineBreakView,
    paragraphBreakView: ParagraphBreakView,
    paragraph: getParagraphComponent,
    link: "a",
    buttonView: FlowButtonView,
    button: "button",
    dynamicTextView: DynamicTextView,
    fallbackView: UnknownNodeView,
});
