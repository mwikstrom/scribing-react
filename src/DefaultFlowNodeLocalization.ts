import { FlowNodeLocalization } from ".";

/** @public */
export const DefaultFlowNodeLocalization: Readonly<FlowNodeLocalization> = Object.freeze({
    holdCtrlKeyToEnableLink: "Hold CTRL key to enable link",
});
