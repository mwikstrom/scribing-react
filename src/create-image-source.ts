import { ImageSource, ImageSourceProps } from "scribing";

/** @public */
export const createImageSource = async (blob: Blob, upload: string): Promise<ImageSource> => {
    const url = URL.createObjectURL(blob);
    try {
        const image = await loadImage(url);
        const scale = getPlaceholderScale(image.width, image.height);
        const canvas = document.createElement("canvas");
        canvas.width = image.width * scale;
        canvas.height = image.height * scale;
        const context = canvas.getContext("2d");
        const props: ImageSourceProps = {
            width: image.width,
            height: image.height,
            url: "",
            upload,
        };
        if (context) {
            context.drawImage(image, 0, 0, canvas.width, canvas.height);
            props.placeholder = getPlaceholderData(canvas);
        }
        return new ImageSource(props);
    } finally {
        URL.revokeObjectURL(url);
    }
};

const loadImage = async (url: string) => new Promise<HTMLImageElement>((resolve, reject) => {
    const image = new Image();
    image.onerror = () => reject(new Error("Failed to load image"));
    image.onload = () => resolve(image);
    image.src = url;
});

const MAX_PLACEHOLDER_PIXELS = 256;
const getPlaceholderScale = (width: number, height: number) => Math.min(
    1, 
    Math.sqrt(MAX_PLACEHOLDER_PIXELS / (width * height))
);

const getPlaceholderData = (canvas: HTMLCanvasElement) => {
    const dataUrl = canvas.toDataURL("image/webp", 0.05);
    const needle = ";base64,";
    const index = dataUrl.indexOf(needle);
    if (index > 0) {
        return dataUrl.substr(index + needle.length);
    }
};
