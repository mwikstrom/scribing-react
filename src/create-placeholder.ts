export const createPlaceholder = (source: HTMLImageElement | HTMLVideoElement): string | undefined => {
    const scale = getPlaceholderScale(source.width, source.height);
    const canvas = document.createElement("canvas");
    canvas.width = source.width * scale;
    canvas.height = source.height * scale;
    const context = canvas.getContext("2d");
    if (context) {
        context.drawImage(source, 0, 0, canvas.width, canvas.height);
        return getPlaceholderData(canvas);
    }
};

const MAX_PLACEHOLDER_PIXELS = 256;
const getPlaceholderScale = (width: number, height: number) => Math.min(
    1, 
    Math.sqrt(MAX_PLACEHOLDER_PIXELS / (width * height))
);

const getPlaceholderData = (canvas: HTMLCanvasElement) => {
    const dataUrl = canvas.toDataURL("video/webp", 0.05);
    const needle = ";base64,";
    const index = dataUrl.indexOf(needle);
    if (index > 0) {
        return dataUrl.substr(index + needle.length);
    }
};
