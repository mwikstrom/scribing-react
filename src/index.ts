/**
 * React components for collaborative rich text editing
 * @packageDocumentation
 */

export * from "./FlowEditor";