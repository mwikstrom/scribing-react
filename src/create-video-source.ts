import { VideoSource, VideoSourceProps } from "scribing";
import { createPlaceholder } from "./create-placeholder";

/** @public */
export function createVideoSource(blob: Blob, upload: string): Promise<VideoSource>;

/** @public */
export function createVideoSource(existingUrl: string): Promise<VideoSource>;

/** @public */
export function createVideoSource(
    blobOrUrl: Blob | string,
    upload?: string
): Promise<VideoSource> {
    if (typeof blobOrUrl === "string") {
        return createVideoSourceFromUrl(blobOrUrl);
    } else if (typeof upload !== "string") {
        throw new TypeError("An upload operation identifier must be specified");
    } else {
        return createUploadVideoSource(blobOrUrl, upload);
    }
}

const createVideoSourceFromUrl = async (url: string): Promise<VideoSource> => {
    const video = await loadVideo(url);
    const placeholder = createPlaceholder(video);
    const props: VideoSourceProps = {
        width: video.width,
        height: video.height,
        url,
    };
    if (placeholder) {
        props.placeholder = placeholder;
    }
    return new VideoSource(props);
};

const createUploadVideoSource = async (blob: Blob, upload: string): Promise<VideoSource> => {
    const url = URL.createObjectURL(blob);
    try {
        const video = await loadVideo(url);
        const props: VideoSourceProps = {
            width: video.width,
            height: video.height,
            url: "",
            upload,
        };
        const placeholder = createPlaceholder(video);
        if (placeholder) {
            props.placeholder = placeholder;
        }
        return new VideoSource(props);
    } finally {
        URL.revokeObjectURL(url);
    }
};

const loadVideo = async (url: string) => new Promise<HTMLVideoElement>((resolve, reject) => {
    const video = document.createElement("video");
    video.onerror = () => reject(new Error("Failed to load video"));
    video.onload = () => resolve(video);
    video.src = url;
});
