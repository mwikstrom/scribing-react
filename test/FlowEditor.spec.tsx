import React from "react";
import renderer from "react-test-renderer";
import { FlowEditor } from "../src";

describe("FlowEditor", () => {
    it("can render without props", () => {
        const component = renderer.create(<FlowEditor/>);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});