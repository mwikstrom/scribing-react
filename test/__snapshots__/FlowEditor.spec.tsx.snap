// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowEditor can render without props 1`] = `<div />`;
